runtime/compiler/env/J9KnownObjectTable.cpp

```diff
@@ -56,28 +65,167 @@ J9::KnownObjectTable::isNull(Index index)
 TR::KnownObjectTable::Index
 J9::KnownObjectTable::getIndex(uintptrj_t objectPointer)
    {
-   TR_J9VMBase *fej9 = (TR_J9VMBase *)(self()->fe());
-   TR_ASSERT(fej9->haveAccess(), "Must haveAccess in J9::KnownObjectTable::getIndex");
    if (objectPointer == 0)
       return 0; // Special Index value for NULL
 
-   // Search for existing matching entry
-   //
    uint32_t nextIndex = self()->getEndIndex();
-   for (uint32_t i = 1; i < nextIndex; i++)
-      if (*_references.element(i) == objectPointer)
-         return i;
-
-   // No luck -- allocate a new one
-   //
-   J9VMThread *thread = getJ9VMThreadFromTR_VM(self()->fe());
-   TR_ASSERT(thread, "assertion failure");
-   _references.setSize(nextIndex+1);
-   _references[nextIndex] = (uintptrj_t*)thread->javaVM->internalVMFunctions->j9jni_createLocalRef((JNIEnv*)thread, (j9object_t)objectPointer);
+#if defined(JITSERVER_SUPPORT)
+   if (self()->comp()->isOutOfProcessCompilation())
+      {
+      auto stream = TR::CompilationInfo::getStream();
+      stream->write(JITServer::MessageType::KnownObjectTable_getIndex, objectPointer);
+      auto recv = stream->read<TR::KnownObjectTable::Index, uintptrj_t *>();
+
+      TR::KnownObjectTable::Index index = std::get<0>(recv);
+      if (index != nextIndex)
+         {
+         return index;
+         }
+      else
+         {
+         _references.setSize(nextIndex+1);
+         _references[nextIndex] = std::get<1>(recv);
+         }
+      }
+   else
+#endif /* defined(JITSERVER_SUPPORT) */
+      {
+      TR_J9VMBase *fej9 = (TR_J9VMBase *)(self()->fe());
+      TR_ASSERT(fej9->haveAccess(), "Must haveAccess in J9::KnownObjectTable::getIndex");
+
+      // Search for existing matching entry
+      //
+      for (uint32_t i = 1; i < nextIndex; i++)
+         if (*_references.element(i) == objectPointer)
+            return i;
+
+      // No luck -- allocate a new one
+      //
+      J9VMThread *thread = getJ9VMThreadFromTR_VM(self()->fe());
+      TR_ASSERT(thread, "assertion failure");
+      _references.setSize(nextIndex+1);
+      _references[nextIndex] = (uintptrj_t*)thread->javaVM->internalVMFunctions->j9jni_createLocalRef((JNIEnv*)thread, (j9object_t)objectPointer);
+      }
+
    return nextIndex;
    }
 
 
+TR::KnownObjectTable::Index
+J9::KnownObjectTable::getIndex(uintptrj_t objectPointer, bool isArrayWithConstantElements)
```

Comment:
Why does this method need to be reimplemented? It should call the correct `getIndex` method, though `addArrayWithConstantElements` also needs to do the right thing (if the client also maintains a KOT, it may need to add the index to the bitvector represented by `_arrayWithConstantElements` though I am unsure).

Reply:
I see a difference. In the original implementation we assume that we have VM access. However, I think that the server always holds VM access.

Reply:
Yes, this method is reimplemented because of the assert that checks on the VM access in the OMR implementation. 
As long as `isArrayWithConstantElements()` is not called at `JITClient`, `_arrayWithConstantElements` doesn’t need to be updated at the `JITClient`. So `addArrayWithConstantElements()` doesn't need to be reimplemented.